apollo-router/src/services/layers/persisted_queries/mod.rs

```diff
@@ -0,0 +1,963 @@
+            ),
+        )
+    }
+
+    /// Create a new disabled [`PersistedQueryLayer`] shutting down the existing manifest poller if it exists.
+    async fn new_disabled(
+        configuration: &Configuration,
+        preexisting_manifest_poller: Option<Arc<PersistedQueryManifestPoller>>,
+    ) -> Result<Self, BoxError> {
+        if let Some(preexisting_manifest_poller) = preexisting_manifest_poller {
+            preexisting_manifest_poller.shutdown().await?;
+        }
+
+        Self::new_with_manifest_poller(configuration, None)
+    }
+
+    fn new_with_manifest_poller(
+        configuration: &Configuration,
+        manifest_poller: Option<Arc<PersistedQueryManifestPoller>>,
+    ) -> Result<Self, BoxError> {
+        Ok(Self {
+            manifest_poller,
+            apq_enabled: configuration.apq.enabled,
+            sandbox_enabled: configuration.sandbox.enabled,
+            safelist_config: configuration.preview_persisted_queries.safelist.clone(),
+            log_unknown: configuration.preview_persisted_queries.log_unknown,
+        })
+    }
+
+    /// Run a request through the layer.
+    /// Takes care of:
+    /// 1) resolving a persisted query ID to a query body
+    /// 2) matching a freeform GraphQL request against persisted queries, optionally rejecting it based on configuration
+    /// 3) continuing to the next stage of the router
+    pub(crate) fn supergraph_request(
+        &self,
+        request: SupergraphRequest,
+    ) -> Result<SupergraphRequest, SupergraphResponse> {
+        if let Some(manifest_poller) = &self.manifest_poller {
+            if let Some(persisted_query_id) = PersistedQueryIdExtractor::extract_id(&request) {
+                self.replace_query_id_with_operation_body(
+                    request,
+                    manifest_poller.clone(),
+                    &persisted_query_id,
+                )
+            } else {
+                self.handle_freeform_graphql(request, manifest_poller.clone())
+            }
+        } else {
+            Ok(request)
+        }
+    }
+
+    /// Places an operation body on a [`SupergraphRequest`] if it has been persisted
+    pub(crate) fn replace_query_id_with_operation_body(
+        &self,
+        mut request: SupergraphRequest,
+        manifest_poller: Arc<PersistedQueryManifestPoller>,
+        persisted_query_id: &str,
+    ) -> Result<SupergraphRequest, SupergraphResponse> {
+        if request.supergraph_request.body().query.is_some() {
+            if self.apq_enabled {
+                // if the request has a query and an ID, and APQ is enabled, continue with normal execution.
+                // safelisting and APQ are incomaptible with each other - therefore we don't need to check
+                // if the ID in the requests exactly maps to the body in the persisted query manifest,
+                // we can just ignore the ID and let APQ handle it for us
+                assert!(!self.safelist_config.enabled);
```

Comment:
I don't understand why there's an assert here, i'd rather have us not do that, and possibly adding an integration test that ensures a router doesn't start up if both are enabled?

Reply:
yeah we do have that test, i added this on a suggestion from @glasser.

Reply:
It definitely seems pretty vital to ensure that later refactorings to how various features interact with each other don't lead to create a giant security hole in the safelist. I suppose just having an integration test might be good enough, but this logic is a bit hairy and the idea of having a codepath that says "I see a body was sent and I'm going to allow it" that doesn't allude to the concept of safelisting seemed scary. Perhaps the entire `supergraph_request` function could be refactored so that the structure is more like

```
if (there is a body) {
  ... do something
} else if (there is an ID) {
  ... do something
}
```

instead of having the top-level logic be "if there is an ID" — that might be a structure that would naturally avoid this assertion.  but in the current structure it really does feel to me that we want to be really really sure that there's no safelist before we say "this document, which we haven't checked against the PQ list, is legit".  (a previous version of this PR did have a bug here where you could effectively circumvent the safelist merely by including a random ID along with a document!)

Reply:
(and I'm not necessarily recommending that restructure as something worth doing)

Reply:
I'm going to raise this as a follow up ticket. We're still in preview, and we have time to change this.

Reply:
Followup sounds good, I'm extremely uncomfortable with asserts in production code (and panics / expects for that matter). we can failfast_debug! and / or return an error, but the router should never ever crash unless it's at startup (it shouldnt crash at reload for example).

I understand this codepath is `unreachable!`, but better safe than sorry.

Reply:
I don't have any strong feelings about *how* returning `Ok(request)` in this case is prevented — returning a 5xx to the user in this hopefully-unreachable case would be fine too. (But also if this didn't block merge/alpha that would be great :) )